modules/diskq/tests/test_qdisk.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2022 One Identity
+ * Copyright (c) 2022 László Várady
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include <criterion/criterion.h>
+
+#include "syslog-ng.h"
+#include "apphook.h"
+#include "qdisk.h"
+#include "scratch-buffers.h"
+
+#include <unistd.h>
+
+/* QDisk-internal: the frame is a 4-byte integer */
+#define FRAME_LENGTH 4
+
+#define DUMMY_RECORD_PATTERN ('z')
+#define MiB(x) (x * 1024 * 1024)
+
+typedef enum
+{
+  TDISKQ_NON_RELIABLE = 0,
+  TDISKQ_RELIABLE = 1,
+} TestDiskQType;
+
+static DiskQueueOptions *
+construct_diskq_options(TestDiskQType dq_type, gint64 disk_buf_size)
+{
+  DiskQueueOptions *opts = g_new0(DiskQueueOptions, 1);
+  disk_queue_options_set_default_options(opts);
+  disk_queue_options_disk_buf_size_set(opts, disk_buf_size);
+  disk_queue_options_reliable_set(opts, dq_type);
+
+  return opts;
+}
+
+static QDisk *
+create_qdisk(const gchar *filename, TestDiskQType dq_type, gint64 disk_buf_size)
+{
+  DiskQueueOptions *opts = construct_diskq_options(TDISKQ_RELIABLE, disk_buf_size);
+  QDisk *qdisk = qdisk_new(opts, "TEST");
+
+  return qdisk;
+}
+
+static void
+cleanup_qdisk(const gchar *filename, QDisk *qdisk)
```

Comment:
I could not, because `qdisk_stop()` deallocates the filename before calling the cleanup function.

I decided not to add `qdisk_stop()` to `cleanup_qdisk()` so that we can play with restarts inside tests.